Add configurable S parameter for 100% power.

madleech opened the pull request:
Different firmwares have different ideas about what `S` values constitutes 100% laser power. E.g. 
* Marlin uses 100
* Smoothie uses 1.0
* Stock Grbl and MrBeam use 1000

This PR adds a setting to the generic gcode driver to make it more versatile at generating gcode for a variety of devices. For Marlin and Smoothie, they have the setting hidden and defaults initialised to the correct values.

The only gotcha here is when loading an existing laser cutter configured in an older version of Visicut, as it won't have the `<spindleMax>` field in its XML. When xstream deserializes it, it ignores any constructor defaults, and even ignores the `protected double spindleMax = 1.0` statement. Instead it initialises this missing field to `0.0` which means your laser power will top out at... zero.

So `setKeysMissingFromDeserialization` is added as a post-deserialisation hook in order to set the missing keys to sensible defaults. On Marlin it sets it to 100, for the others it is set to 1.0.

Example gcode, the usual two 20mm boxes, one at 20%, one at 100% laser strength.

**Generic gcode driver**:
```
G21
G90
M3
G0 X0.000000 Y0.000000 F3600
G1 X19.964400 Y0.000000 S0.200000 F1200
G1 X19.964400 Y19.964400
G1 X0.000000 Y19.964400
G1 X0.000000 Y0.000000
G1 X0.000000 Y0.000000
G0 X24.993600 Y0.000000 F3600
G1 X44.958000 Y0.000000 S1.000000 F1200
G1 X44.958000 Y19.964400
G1 X24.993600 Y19.964400
G1 X24.993600 Y0.000000
G1 X24.993600 Y0.000000
G0 X0 Y0
M5
```

**Generic gcode with 1000**:
```
G21
G90
M3 S0
G0 X0.000000 Y0.000000 F3600
G1 X19.964400 Y0.000000 S250.000000 F400
G1 X19.964400 Y19.964400
G1 X0.000000 Y19.964400
G1 X0.000000 Y0.000000
G1 X0.000000 Y0.000000
G0 X24.993600 Y0.000000 F3600
G1 X44.958000 Y0.000000 S1000.000000 F500
G1 X44.958000 Y19.964400
G1 X24.993600 Y19.964400
G1 X24.993600 Y0.000000
G1 X24.993600 Y0.000000
M5
G0 X0 Y150
```

**Marlin**:
```
G21
G90
G28 XY
M5
G0 X0.000000 Y0.000000 F3600
G1 X19.964400 Y0.000000 S20.000000 F1200
G1 X19.964400 Y19.964400
G1 X0.000000 Y19.964400
G1 X0.000000 Y0.000000
G1 X0.000000 Y0.000000
G0 X24.993600 Y0.000000 F3600
G1 X44.958000 Y0.000000 S100.000000 F1200
G1 X44.958000 Y19.964400
G1 X24.993600 Y19.964400
G1 X24.993600 Y0.000000
G1 X24.993600 Y0.000000
G0 X0 Y0
M5
G28 XY
```

**Smoothie**:
```
G21
G90
M3
G0 X0.000000 Y0.000000 F3600
G1 X19.964400 Y0.000000 S0.200000 F1200
G1 X19.964400 Y19.964400
G1 X0.000000 Y19.964400
G1 X0.000000 Y0.000000
G1 X0.000000 Y0.000000
G0 X24.993600 Y0.000000 F3600
G1 X44.958000 Y0.000000 S1.000000 F1200
G1 X44.958000 Y19.964400
G1 X24.993600 Y19.964400
G1 X24.993600 Y0.000000
G1 X24.993600 Y0.000000
G0 X0 Y0
M5
```

Does anyone have a better suggestion for the settings label? Currently "100% value for spindle" which is obviously not very good. "Max S value for laser"? "Max RPM value for laser-spindle"?

mgmax commented:
First, a big thumbs up for the quality of your pull requests! While I didn't find time to test your code (and am not personally affected by it), both your code contributions and pull requests are very nice to review.

Regarding the settings label, I would focus on the technical terms and not the interpretation: "S value for 100% laser power". If someone doesn't understand this, he most probably also won't understand any other description.

mgmax commented:
and by the way @madleech, greetings to the other end of the world, I was in CHC a few months ago but somehow didn't find time to visit the FabLab. What a strange small world...

madleech commented:
Thanks @mgmax, small world indeed :) I'm not in CHC any more, moved north to a very much smaller town, but we have a _free_ MakerCrate (i.e. a container filled with laser cutters and 3d printers) in the centre of town!

Updated the setting label, sounds good and I agree with your reasoning.

t-oster commented:
@madleech please rebase

t-oster commented:
And I second @mgmax, it is really a breeze of fresh air to review and merge your pull-requests. We need more developers like you.

madleech commented:
Thanks @t-oster, Visicut is a really great program, it's a pleasure to work on it! Branch rebased.
